source/adapters/opencl/device.hpp

```diff
@@ -10,11 +10,97 @@
 #pragma once
 
 #include "common.hpp"
+#include "platform.hpp"
 
-namespace cl_adapter {
-ur_result_t getDeviceVersion(cl_device_id Dev, oclv::OpenCLVersion &Version);
+struct ur_device_handle_t_ {
+  using native_type = cl_device_id;
+  native_type Device;
+  ur_platform_handle_t Platform;
+  cl_device_type Type = 0;
+  ur_device_handle_t ParentDevice = nullptr;
+  std::atomic<uint32_t> RefCount = 0;
+  bool IsNativeHandleOwned = true;
 
-ur_result_t checkDeviceExtensions(cl_device_id Dev,
-                                  const std::vector<std::string> &Exts,
-                                  bool &Supported);
-} // namespace cl_adapter
+  ur_device_handle_t_(native_type Dev, ur_platform_handle_t Plat,
+                      ur_device_handle_t Parent)
+      : Device(Dev), Platform(Plat), ParentDevice(Parent) {
+    RefCount = 1;
+    if (Parent) {
+      Type = Parent->Type;
+    } else {
+      clGetDeviceInfo(Device, CL_DEVICE_TYPE, sizeof(cl_device_type), &Type,
+                      nullptr);
+    }
+  }
+
+  ~ur_device_handle_t_() {
+    if (ParentDevice && IsNativeHandleOwned) {
+      clReleaseDevice(Device);
+    }
+  }
+
+  uint32_t incrementReferenceCount() noexcept { return ++RefCount; }
+
+  uint32_t decrementReferenceCount() noexcept { return --RefCount; }
+
+  uint32_t getReferenceCount() const noexcept { return RefCount; }
+
+  native_type get() { return Device; }
+
+  ur_result_t getDeviceVersion(oclv::OpenCLVersion &Version) {
+    size_t DevVerSize = 0;
+    CL_RETURN_ON_FAILURE(
+        clGetDeviceInfo(Device, CL_DEVICE_VERSION, 0, nullptr, &DevVerSize));
+
+    std::string DevVer(DevVerSize, '\0');
+    CL_RETURN_ON_FAILURE(clGetDeviceInfo(Device, CL_DEVICE_VERSION, DevVerSize,
+                                         DevVer.data(), nullptr));
+
+    Version = oclv::OpenCLVersion(DevVer);
+    if (!Version.isValid()) {
+      return UR_RESULT_ERROR_INVALID_DEVICE;
+    }
+
+    return UR_RESULT_SUCCESS;
+  }
+
+  bool isIntelFPGAEmuDevice() {
+    size_t NameSize = 0;
+    CL_RETURN_ON_FAILURE(
+        clGetDeviceInfo(Device, CL_DEVICE_NAME, 0, nullptr, &NameSize));
+    std::string NameStr(NameSize, '\0');
+    CL_RETURN_ON_FAILURE(clGetDeviceInfo(Device, CL_DEVICE_NAME, NameSize,
+                                         NameStr.data(), nullptr));
+
+    return NameStr.find("Intel(R) FPGA Emulation Device") != std::string::npos;
+  }
+
+  ur_result_t checkDeviceExtensions(const std::vector<std::string> &Exts,
+                                    bool &Supported) {
```

Comment:
This interface isn't very nice to use. Do we really need to return the `ur_result_t` here? Could it just return `false`?

This is another issue that can be looked at separately.

Reply:
Didn't look into this area much but I think this could be addressed in a follow-up that would redesign these extensions helper functions.

Reply:
Makes sense.